presto-memory-context/src/main/java/com/facebook/presto/memory/MemoryTrackingContext.java

```diff
@@ -0,0 +1,185 @@
+ * updateUserMemory()/updateRevocableMemory()/updateSystemMemory() methods.
+ */
+@ThreadSafe
+public class MemoryTrackingContext
+{
+    private final AggregatedMemoryContext userAggregateMemoryContext;
+    private final AggregatedMemoryContext revocableAggregateMemoryContext;
+    private final AggregatedMemoryContext systemAggregateMemoryContext;
+
+    private final LocalMemoryContext userLocalMemoryContext;
+    private final LocalMemoryContext revocableLocalMemoryContext;
+    private final LocalMemoryContext systemLocalMemoryContext;
+
+    private final AtomicLong peakUserMemory = new AtomicLong();
+
+    public MemoryTrackingContext(
+            AggregatedMemoryContext userAggregateMemoryContext,
+            AggregatedMemoryContext revocableAggregateMemoryContext,
+            AggregatedMemoryContext systemAggregateMemoryContext)
+    {
+        this.userAggregateMemoryContext = requireNonNull(userAggregateMemoryContext, "userAggregateMemoryContext is null");
+        this.revocableAggregateMemoryContext = requireNonNull(revocableAggregateMemoryContext, "revocableAggregateMemoryContext is null");
+        this.systemAggregateMemoryContext = requireNonNull(systemAggregateMemoryContext, "systemAggregateMemoryContext is null");
+        this.userLocalMemoryContext = userAggregateMemoryContext.newLocalMemoryContext();
+        this.revocableLocalMemoryContext = revocableAggregateMemoryContext.newLocalMemoryContext();
+        this.systemLocalMemoryContext = systemAggregateMemoryContext.newLocalMemoryContext();
+    }
+
+    public ListenableFuture<?> updateUserMemory(long delta)
+    {
+        if (delta >= 0) {
+            ListenableFuture<?> future = userLocalMemoryContext.addBytes(delta);
+            peakUserMemory.accumulateAndGet(userLocalMemoryContext.getBytes(), Math::max);
+            return future;
+        }
+        userLocalMemoryContext.addBytes(delta);
+        return NOT_BLOCKED;
+    }
+
+    public boolean tryReserveUserMemory(long delta)
+    {
+        if (userLocalMemoryContext.tryReserveMemory(delta)) {
+            peakUserMemory.accumulateAndGet(userLocalMemoryContext.getBytes(), Math::max);
+            return true;
+        }
+        return false;
+    }
+
+    public ListenableFuture<?> updateRevocableMemory(long delta)
+    {
+        if (delta >= 0) {
+            return revocableLocalMemoryContext.addBytes(delta);
+        }
+        revocableLocalMemoryContext.addBytes(delta);
+        return NOT_BLOCKED;
+    }
+
+    public void close()
+    {
+        try (Closer closer = Closer.create()) {
+            closer.register(userAggregateMemoryContext::close);
+            closer.register(revocableAggregateMemoryContext::close);
+            closer.register(systemAggregateMemoryContext::close);
+            userLocalMemoryContext.setBytes(0);
+            revocableLocalMemoryContext.setBytes(0);
+            systemLocalMemoryContext.setBytes(0);
```

Comment:
Explain why local memory doesn't go in closer, whereas aggregate memory does.

In addition, I vaguely remember that you intended to add `close` method to local memory context due to address race issues.